Source/Matter/NewMatterConstraints.impl.hpp

```diff
@@ -0,0 +1,66 @@
+/* GRChombo
+ * Copyright 2012 The GRChombo collaboration.
+ * Please refer to LICENSE in GRChombo's root directory.
+ */
+
+#if !defined(NEWMATTERCONSTRAINTS_HPP_)
+#error "This file should only be included through NewMatterConstraints.hpp"
+#endif
+
+#ifndef NEWMATTERCONSTRAINTS_IMPL_HPP_
+#define NEWMATTERCONSTRAINTS_IMPL_HPP_
+#include "DimensionDefinitions.hpp"
+
+template <class matter_t>
+MatterConstraints<matter_t>::MatterConstraints(
+    const matter_t a_matter, double dx, int a_c_Ham, const Interval &a_c_Moms,
+    int a_c_Ham_abs_terms /* defaulted*/,
+    const Interval &a_c_Moms_abs_terms /*defaulted*/,
+    double G_Newton /*defaulted*/)
+    : Constraints(dx, a_c_Ham, a_c_Moms, a_c_Ham_abs_terms, a_c_Moms_abs_terms,
+                  0.0 /*No cosmological constant*/),
+      my_matter(a_matter), m_G_Newton(G_Newton)
+{
+}
+
+template <class matter_t>
+template <class data_t>
+void MatterConstraints<matter_t>::compute(Cell<data_t> current_cell) const
+{
+    // Load local vars and calculate derivs
+    const auto vars = current_cell.template load_vars<BSSNMatterVars>();
+    const auto d1 = m_deriv.template diff1<BSSNMatterVars>(current_cell);
+    const auto d2 = m_deriv.template diff2<BSSNMatterVars>(current_cell);
+
+    // Get the non matter terms for the constraints
+    Vars<data_t> out = constraint_equations(vars, d1, d2);
+
+    // Inverse metric and Christoffel symbol
+    const auto h_UU = TensorAlgebra::compute_inverse_sym(vars.h);
+    const auto chris = TensorAlgebra::compute_christoffel(d1.h, h_UU);
```

Comment:
Here I would suggest the following:
`constraint_equations` is only really used internally. But inside it we calculate again `h_UU` and `chris`
I would much prefer if these were calculated once and passed as arguments by reference to `constraint_equations` like the derivatives.
What do you think?

Reply:
I agree that this makes sense. Do you think it's safe to change the old `Constraints` class as well?

Reply:
Agreed, and yes I think this should be ok in the old one too.

Reply:
I think some people might have changed their MatterConstraints class to save rho and stuff like that, so it might generate conflicts depending on how they did it, but I would be ok with it (as in, I would like to see that conflict and realize I could make that change, because to be honest I don't have it in my code and would prefer if it was that way). So I would say yes as well, though not very important. But on the new classes yes